packages/hardhat-viem/src/internal/tasks.ts

```diff
@@ -0,0 +1,302 @@
+  const fqn = getFullyQualifiedName(sourceName, contractName);
+  const validNames = isDuplicate ? [fqn] : [contractName, fqn];
+  const json = JSON.stringify(artifact, undefined, 2);
+  const contractTypeName = `${contractName}$Type`;
+
+  const constructorAbi = artifact.abi.find(
+    ({ type }) => type === "constructor"
+  );
+
+  const inputs: Array<{
+    internalType: string;
+    name: string;
+    type: string;
+  }> = constructorAbi !== undefined ? constructorAbi.inputs : [];
+
+  const constructorArgs =
+    inputs.length > 0
+      ? `constructorArgs: [${inputs
+          .map(({ name, type }) => getArgType(name, type))
+          .join(", ")}]`
+      : `constructorArgs?: []`;
+
+  return `${AUTOGENERATED_FILE_PREFACE}
+
+import type { Address } from "viem";
+${
+  inputs.length > 0
+    ? `import type { AbiParameterToPrimitiveType, GetContractReturnType } from "@nomicfoundation/hardhat-viem/types";`
+    : `import type { GetContractReturnType } from "@nomicfoundation/hardhat-viem/types";`
+}
+import "@nomicfoundation/hardhat-viem/types";
+
+export interface ${contractTypeName} ${json}
+
+declare module "@nomicfoundation/hardhat-viem/types" {
+  ${validNames
+    .map(
+      (name) => `export function deployContract(
+    contractName: "${name}",
+    ${constructorArgs},
+    config?: DeployContractConfig
+  ): Promise<GetContractReturnType<${contractTypeName}["abi"]>>;`
+    )
+    .join("\n  ")}
+
+  ${validNames
+    .map(
+      (name) => `export function getContractAt(
+    contractName: "${name}",
+    address: Address,
+    config?: GetContractAtConfig
+  ): Promise<GetContractReturnType<${contractTypeName}["abi"]>>;`
+    )
+    .join("\n  ")}
+}
+`;
+}
+
+/**
+ * Generates TypeScript code to extend the `ArtifactsMap` interface with
+ * contract types.
+ */
+function generateArtifactsDefinition(
+  contractTypeData: Array<{
+    contractName: string;
+    fqn: string;
+    typeName: string;
+    declaration: string;
+  }>
+) {
+  return `${AUTOGENERATED_FILE_PREFACE}
+
+import "hardhat/types/artifacts";
+
+${contractTypeData
+  .map((ctd) => `import { ${ctd.typeName} } from "./${ctd.contractName}";`)
+  .join("\n")}
+
+declare module "hardhat/types/artifacts" {
+  interface ArtifactsMap {
+    ${contractTypeData
+      .map((ctd) => `["${ctd.contractName}"]: ${ctd.typeName};`)
+      .join("\n    ")}
+    ${contractTypeData
+      .map((ctd) => `["${ctd.fqn}"]: ${ctd.typeName};`)
+      .join("\n    ")}
+  }
+}
+`;
+}
+
+/**
+ * Returns the type of a function argument in one of the following formats:
+ * - If the 'name' is provided:
+ *   "name: AbiParameterToPrimitiveType<{ name: string; type: string; }>"
+ *
+ * - If the 'name' is empty:
+ *   "AbiParameterToPrimitiveType<{ name: string; type: string; }>"
+ */
+function getArgType(name: string, type: string) {
+  const argType = `AbiParameterToPrimitiveType<${JSON.stringify({
+    name,
+    type,
+  })}>`;
+
+  return name !== "" ? `${name}: ${argType}` : argType;
```

Comment:
How sure are we that `name` cannot be `undefined` here? It comes from casting an `any` read from a json file, so I'm not 100% sure that it's impossible, but maybe the ABI output is specified in a way that makes `name` always a string.

Reply:
`abitype` at least seems to think it can be `undefined`: https://github.com/frangio/abitype/blob/3879b593580d3fecc0d585d82f4876c7e6bd8d7b/src/abi.ts#L101

Reply:
Solc v0.5.1, which is the earliest version [we support](https://hardhat.org/hardhat-runner/docs/reference/solidity-support), generates empty strings for anonymous parameters. So idk. I think I would be defensive anyway just in case, but feel free to leave it as it is if you think it's unnecessary.